File: leet_code/javascript_30/javascript/2665.js
/**
 * @param {integer} init
 * @return { increment: Function, decrement: Function, reset: Function }
 */
var createCounter = function(init) {
    let lastVal = init;
    return {
        increment: () => {
            return lastVal += 1;
        },
        decrement: () => {
            return lastVal -= 1;
        },
        reset: () => {
            return lastVal = init;
        }
    }
};

/**
 * const counter = createCounter(5)
 * counter.increment(); // 6
 * counter.reset(); // 5
 * counter.decrement(); // 4
 */

File: leet_code/javascript_30/go/2665.go

package main

import "fmt"

type Counter struct {
    init    int
    lastVal int
}

func (c *Counter) Increment() int {
    c.lastVal++
    return c.lastVal
}

func (c *Counter) Decrement() int {
    c.lastVal--
    return c.lastVal
}

func (c *Counter) Reset() int {
    c.lastVal = c.init
    return c.lastVal
}

func createCounter(init int) *Counter {
    return &Counter{init: init, lastVal: init}
}

func main() {
    counter := createCounter(5)
    fmt.Println(counter.Increment()) // 6
    fmt.Println(counter.Decrement()) // 5
    fmt.Println(counter.Reset())     // 5
}
